src/blog/2024/01/dashboard-2-multi-user.md

```diff
@@ -0,0 +1,127 @@
+<figcaption class="-mt-6 mb-4 text-center"><b>The "Admin" view that is only made available to users registered as an "admin".</b></figcaption>
+
+Under the covers, we're appending our `user` object to the `msg` object, via the SocketIO `auth` option. We make the `socketio` object available via a computed `setup` variable, this means that we can access this data in any `ui-template` node, and render like so:
+
+{% raw %}
+```html
+<template>
+    <div>
+        <h1>Hi, {{ setup?.socketio?.auth?.user?.name }}!</h1>
+    </div>
+</template>
+```
+{% endraw %}
+
+To enable custom user-by-user content in a `ui-template` though, we must allow it to "Accept Client Constraints". This means that if a `.msg._client.user` value is included in any messages sent to a `ui-template` node, then the underlying SocketIO message will be filtered to only send to the relevant user's connection, and no others.
+
+### Admin Only Views
+
+With this new functionality we can also now show/hide content based on the authenticated user. 
+
+We recently introduced the option to [set default visiblity & interaction states](https://github.com/FlowFuse/node-red-dashboard/pull/484). This was partly introduced because it's a good practice to set the default "Visibility" option for any admin-only pages to "Hidden", and then use a `ui-control` node to show the content only to the relevant admins.
+
+<iframe width="100%" height="100%" src="https://flows.nodered.org/flow/2fe8e6f1e7002f1ff6a9195ad1a153b6/share" allow="clipboard-read; clipboard-write" style="border: none; margin-bottom: 12px;"></iframe>
+
+Let's breakdown the above flow:
+
+1. We wire a `ui-event` node (which emits each time a user views a page) into a switch node
+2. Our switch node checks the `user.username` against a known list of admin users and branches "admin:" and "non-admin" users
+3. For admin users, the `change` node defines a message for our `ui-control` node to dynamically show content, in this case an "Admin" page, when appropriate.
+
+```json
+{
+    "pages":{
+        "show": ["Admin View"]
+    }
+}
+```
+
+All events going into `ui-control` are automatically filtered based on the `msg._client.user` object, so only the Admin users will receive the message to show the "Admin View" page, resulting in:
+
+![Showing Admin Only View](./images/multi-user-dashboard-admin.png "Showing Admin Only View"){data-zoomable}
+<figcaption class="-mt-6 mb-4 text-center"><b>The "Admin" view that is only made available to users registered as an "admin".</b></figcaption>
+
+Further extensions of this could also check `ui-event` in case a non-admin user tries to access the `/admin` page directly, in which case we can utilise `ui-control` to navigate the user away from the page immediately. See the [ui-control documentation](https://dashboard.flowfuse.com/nodes/widgets/ui-control.html#navigation) for more details on this.
```

Comment:
```suggestion
Further extensions of this could also check `ui-event` in case a non-admin user tries to access the `/admin` page directly, in which case we can utilize `ui-control` to navigate the user away from the page immediately. See the [ui-control documentation](https://dashboard.flowfuse.com/nodes/widgets/ui-control.html#navigation) for more details on this.
```

Reply:
https://flowfuse.com/handbook/customer/marketing/website/#marketing-website

All content is written in UK English 😄